src/metabase/models/params/card_values.clj

```diff
@@ -0,0 +1,52 @@
+(ns metabase.models.params.card-values
+  "Code related to getting values for a parameter where its source values is a card."
+  (:require
+    [clojure.string :as str]
+    [metabase.models :refer [Card]]
+    [metabase.query-processor :as qp]
+    [metabase.util.schema :as su]
+    [schema.core :as s]
+    [toucan.db :as db]))
+
+(def ^:dynamic *max-rows*
+  "Maximum number of rows returned when running a card."
+  2000)
+
+(defn- values-from-card-query
+  [card-id value-field query]
+  (let [card        (db/select-one Card :id card-id)]
+    {:database (:database_id card)
+     :type     :query
+     :query    (merge
+                 {:source-table (format "card__%d" card-id)
+                  :fields       [value-field]
+                  :limit        *max-rows*}
+                 (when query
+                   {:filter [:contains [:lower value-field] (str/lower-case query)]}))
+     :middleware {:disable-remaps? true}}))
+
+(s/defn values-from-card
+  "Get values a field from a card.
+
+  (values-from-card 1 [:field \"name\" nil] \"red\")
+  ;; will execute a mbql that looks like
+  ;; {:source-table (format \"card__%d\" card-id)
+  ;;  :fields       [value-field]
+  ;;  :filter       [:contains [:lower value-field] \"red\"]
+  ;;  :limit        *max-rows*}
+  =>
+  {:values          [\"Red Medicine\"]
+   :has_more_values false}
+  "
+  ([card-id value-field-ref]
+   (values-from-card card-id value-field-ref nil))
+
+  ([card-id     :- su/IntGreaterThanZero
+    value-field :- su/Field
```

Comment:
The schema for `Field` is too loose. I would validate here that the Field is only what is necessary to select a field, and excludes binning etc

Reply:
We're using the same schema on MBQL. And this field is used for the same purpose of building the `:fields` in MBQL, so I don't think we need to be to strict about it.

https://github.com/metabase/metabase/blob/master/shared/src/metabase/mbql/schema.cljc#L1370

Reply:
It would be nice to know that value-field doesn't contain any options like `:temporal-unit`,  `:source-field`, or `:join-alias`. It means that it can be safely used in `values-from-card-query` like you're doing without running into issues.